core/src/main/java/org/apache/iceberg/DefaultValueParser.java

```diff
@@ -0,0 +1,374 @@
+        generator.writeString(DateTimeUtil.daysToIsoDate((Integer) defaultValue));
+        break;
+      case TIME:
+        Preconditions.checkArgument(
+            defaultValue instanceof Long, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeString(DateTimeUtil.microsToIsoTime((Long) defaultValue));
+        break;
+      case TIMESTAMP:
+        Preconditions.checkArgument(
+            defaultValue instanceof Long, "Invalid default %s value: %s", type, defaultValue);
+        if (((Types.TimestampType) type).shouldAdjustToUTC()) {
+          generator.writeString(DateTimeUtil.microsToIsoDateTimeTz((Long) defaultValue));
+        } else {
+          generator.writeString(DateTimeUtil.microsToIsoDateTime((Long) defaultValue));
+        }
+        break;
+      case STRING:
+        Preconditions.checkArgument(
+            defaultValue instanceof CharSequence, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeString(((CharSequence) defaultValue).toString());
+        break;
+      case UUID:
+        Preconditions.checkArgument(defaultValue instanceof UUID, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeString(defaultValue.toString());
+        break;
+      case FIXED:
+        Preconditions.checkArgument(
+            defaultValue instanceof ByteBuffer, "Invalid default %s value: %s", type, defaultValue);
+        ByteBuffer byteBufferDefaultValue = (ByteBuffer) defaultValue;
+        Preconditions.checkArgument(
+            byteBufferDefaultValue.hasArray() && byteBufferDefaultValue.arrayOffset() == 0 &&
+                byteBufferDefaultValue.position() == 0,
+            "Invalid default %s value: %s, not a valid bytebuffer representation",
+            type, defaultValue);
+        int actualLength = byteBufferDefaultValue.remaining();
+        int expectedLength = ((Types.FixedType) type).length();
+        Preconditions.checkArgument(actualLength == expectedLength,
+            "Invalid default %s value: %s, incorrect length: %s",
+            type, defaultValue, actualLength);
+        String fixedString = BaseEncoding.base16().encode(ByteBuffers.toByteArray(((ByteBuffer) defaultValue)));
+        generator.writeString(fixedString);
+        break;
+      case BINARY:
+        Preconditions.checkArgument(
+            defaultValue instanceof ByteBuffer, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeString(BaseEncoding.base16().encode(ByteBuffers.toByteArray(((ByteBuffer) defaultValue))));
+        break;
+      case DECIMAL:
+        Preconditions.checkArgument(
+            defaultValue instanceof BigDecimal &&
+                ((BigDecimal) defaultValue).scale() == ((Types.DecimalType) type).scale(),
+            "Invalid default %s value: %s", type, defaultValue);
+        generator.writeString(((BigDecimal) defaultValue).toPlainString());
```

Comment:
From the documentation for `toPlainString`, I think this is a problem:

> if this BigDecimal has a negative scale, the string resulting from this method will have a scale of zero when processed by the string constructor.

That means there's still a case where round-trip conversion will fail. We have two choices:
* Document this in the spec and use the scale of the type (don't fail if the type scale is negative and the number's scale is 0)
* Use a different representation that encodes the scale, like `toString` that is guaranteed to round-trip correctly

I would probably opt for the result of `toPlainString` when the scale is positive and `toString` when the scale is negative.

Reply:
We're also going to need to add more test cases for this.

Reply:
done